### ISSUE TITLE
Ruby 2.2.2 build fail error on macOS 11.1

### ISSUE THREAD

**Issue by laynejohnson** [rbenv/ruby-build#1700]
Hello there! (And, Happy New Year!) I hope I am posting this in the right place! 🙏

I am trying to install Ruby 2.2.2 on macOS 11.1 for a project, but am running into the following error when running `rbenv install 2.2.2` :

`Downloading openssl-1.0.2u.tar.gz...
-> https://dqw8nmjcqpjn7.cloudfront.net/ecd0c6ffb493dd06707d38b14bb4d8c2288bb7033735606569d8f90f89669d16
Installing openssl-1.0.2u...
Installed openssl-1.0.2u to /Users/findum/.rbenv/versions/2.2.2

Downloading ruby-2.2.2.tar.bz2...
-> https://cache.ruby-lang.org/pub/ruby/2.2/ruby-2.2.2.tar.bz2
Installing ruby-2.2.2...

WARNING: ruby-2.2.2 is past its end of life and is now unsupported.
It no longer receives bug fixes or critical security updates.

ruby-build: using readline from homebrew

BUILD FAILED (macOS 11.1 using ruby-build 20201225)

Inspect or clean up the working tree at /var/folders/x8/g8yqvrh55gq0wds4hny1t94r0000gn/T/ruby-build.20210101111906.42540.uFRvAN
Results logged to /var/folders/x8/g8yqvrh55gq0wds4hny1t94r0000gn/T/ruby-build.20210101111906.42540.log

Last 10 log lines:
compiling ../.././ext/psych/yaml/parser.c
linking shared-object json/ext/parser.bundle
linking shared-object pathname.bundle
installing default psych libraries
linking shared-object json/ext/generator.bundle
linking shared-object bigdecimal.bundle
linking shared-object psych.bundle
linking shared-object nkf.bundle
linking shared-object date_core.bundle
make: *** [build-ext] Error 2`

I have tried the workarounds in issue #1353 and have downloaded OpenSSL 1.0 on my system, but still no luck! Has anyone had a similar issue that they were able to resolve? Many thanks 🙏

**Comment by rahulpuroht** [comment]
Having same issue with 2.3.3.Have you found any workaround yet.

**Comment by laynejohnson** [comment]
@rahulpuroht Yes! It seems to have been an issue with XCode 12 (Apple changed the default `CFLAGS`). This solution worked for me: https://github.com/rbenv/ruby-build/issues/1489#issue-741155490. I hope you have some luck with it!

**Comment by eveevans** [comment]
Having the same problem on Mac M1
And the fix didn't works for me... 
also trying using `arch -x86_64 /bin/bash -c 'rbenv install 2.3.8'`
and `RUBY_CFLAGS="-w" arch -x86_64 /bin/bash -c 'rbenv install 2.3.8'`

:(

**Comment by stevenjohn** [comment]
Hi there,

Yes - also trying to get Ruby installed on Mac Silicone m1.

What I've found so far...

`Steves-MacBook-Air:~ steve$ arch -x86_64 /bin/bash -c 'CFLAGS="-Wno-error=implicit-function-declaration" rbenv install 2.2.5'
Installing openssl-1.0.2u...
Installed openssl-1.0.2u to /Users/steve/.rbenv/versions/2.2.5

Installing ruby-2.2.5...

WARNING: ruby-2.2.5 is past its end of life and is now unsupported.
It no longer receives bug fixes or critical security updates.

ruby-build: using readline from homebrew

BUILD FAILED (macOS 11.1 using ruby-build 20201225)

Inspect or clean up the working tree at /var/folders/4v/zps5vz_n3xjdmvq8pl_lf9780000gn/T/ruby-build.20210116112712.11791.4koc3o
Results logged to /var/folders/4v/zps5vz_n3xjdmvq8pl_lf9780000gn/T/ruby-build.20210116112712.11791.log

Last 10 log lines:
#define HAVE_BN_GENCB defined(HAVE_RSA_GENERATE_KEY_EX) || defined(HAVE_DH_GENERATE_PARAMETERS_EX) || defined(HAVE_DSA_GENERATE_PARAMETERS_EX)
                                                           ^
./ossl_pkey.h:44:5: warning: macro expansion producing 'defined' has undefined behavior [-Wexpansion-to-defined]
./ossl_pkey.h:43:103: note: expanded from macro 'HAVE_BN_GENCB'
#define HAVE_BN_GENCB defined(HAVE_RSA_GENERATE_KEY_EX) || defined(HAVE_DH_GENERATE_PARAMETERS_EX) || defined(HAVE_DSA_GENERATE_PARAMETERS_EX)
                                                                                                      ^
3 warnings generated.
3 warnings generated.
linking shared-object openssl.bundle
make: *** [build-ext] Error 2
`

Using:
arch -x86_64 /bin/bash -c
.... has allowed me to get openssl 1.0.2 to install correctly (without this it fails installing ssl)

and I added: 
CFLAGS="-Wno-error=implicit-function-declaration"
.... thanks to this post https://github.com/rbenv/ruby-build/issues/1489

But as you see its not working.

My next attempt followed this: https://stackoverflow.com/questions/64029645/rbenv-ruby-build-installs-fail-for-ruby-versions-1-9-3-2-2-0-2-2-5-with-xcod

which suggests installing command line tools v11, and the setting that in Xcode preferences.

However, on downloading the latest CLTs for v11 (11.5) - the installation says it is not compatible with m1 silicon.

Has anyone got older Ruby with rbenv working on m1 yet? Or found any other steps forward?

Many thanks, Steve.

**Comment by hsbt** [comment]
> Has anyone got older Ruby with rbenv working on m1 yet?

There is no way to build the old ruby at M1 macOS with arm arch.

**Comment by stevenjohn** [comment]
> > Has anyone got older Ruby with rbenv working on m1 yet?
> 
> There is no way to build the old ruby at M1 macOS with arm arch.

I'm sure there'll be a way - else there'll be a very upset group of devs not able to work on older ruby versions.

**Comment by hsbt** [comment]
@stevenjohn Maybe, the old Rubies can work under the Rosseta2 environment. But I have only Apple DTK, I couldn't investigate macOS with M1 chip.

**Comment by stevenjohn** [comment]
Cheers @hsbt - I've tried using 

`arch -x86_64 /bin/bash -c`

as a prefix which I believe uses Rosseta2

By using this I can see my installation went lot further than before - but I think the issue now is with the v12 command line tools, I'm seeing a lot of references to needing to use v11 - but not sure these can be installed on m1.

**Comment by eveevans** [comment]
I know that this is a rbenv post. But I want to comment that I was able to install an old Ruby (2.3.8) on my M1 with Big Sur 11.1, running XCode 12.3 with the following commands: 

```
rvm pkg install openssl
export optflags="-Wno-error=implicit-function-declaration"
rvm install ruby-2.3.8 --with-openssl-dir=$HOME/.rvm/usr --rubygems ignore
```

1. This install old openSSL as independent dependency (not used for the system) 
2. Change a flag that (I think) was changed on XCode 12
3. Install  Ruby using the openSSL dependency from RVM (not from the system)

**Comment by stevenjohn** [comment]
> I know that this is a rbenv post. But I want to comment that I was able to install an old Ruby (2.3.8) on my M1 with Big Sur 11.1, running XCode 12.3 with RVM using the following commands:

That's greatly appreciated - I will try it out and report back.

**Comment by mrmanc** [comment]
It isn’t exactly the same issue, but I wanted to share that the [solution](https://github.com/rbenv/ruby-build/issues/1700#issuecomment-761864381) shared by @eveevans which set `optflags` solved the problem I was having running `rbenv install 2.7.1`. Setting `CFLAGS` did not work for me.

**Comment by ThomasKoppensteiner** [comment]
Thank you @eveevans and @mrmanc the following worked for me:

```sh
export optflags="-Wno-error=implicit-function-declaration"
rbenv install 2.1.2
```

**Comment by bradical** [comment]
Is there a consensus here? I'm seeing the following error on install:

```
config.status: creating -darwin20-fake.rb
rm: illegal option -- a
usage: rm [-f | -i] [-dPRrvW] file ...
       unlink file
config.status: error: could not create -darwin20-fake.rb
make: *** [-darwin20-fake.rb] Error 64
```

after trying https://github.com/rbenv/ruby-build/issues/1700#issuecomment-799541819

```
brew install readline openssl rbenv # note this installs openssl@1.1 for me

# add the following to ~/.zshrc
eval "$(rbenv init -)"
# rbenv install recommends this with ruby-build
export RUBY_CONFIGURE_OPTS="--with-openssl-dir=$(brew --prefix openssl@1.1)"

rbenv install 2.2.2
```

Are we just not able to build this version of ruby for ARM?

**Comment by eveevans** [comment]
@bradical in my case, I ended using Old rubies compiled with Rosseta 2, and the newest Rubies compiled for ARM. 
(with Asdf)

**Comment by bradical** [comment]
Thanks @eveevans!

This is interesting. It seems to have gotten pretty far:

```
make[2]: Nothing to be done for `all'.
make[2]: Nothing to be done for `all'.
make[2]: Nothing to be done for `all'.
make[2]: Nothing to be done for `all'.
make[2]: `ruby' is up to date.
Generating RDoc documentation

No newer files.

  Files:      0

  Classes:    0 (0 undocumented)
  Modules:    0 (0 undocumented)
  Constants:  0 (0 undocumented)
  Attributes: 0 (0 undocumented)
  Methods:    0 (0 undocumented)

  Total:      0 (0 undocumented)
    0.00% documented

  Elapsed: 0.0s

config.status: creating -darwin20-fake.rb
rm: illegal option -- a
usage: rm [-f | -i] [-dPRrvW] file ...
       unlink file
config.status: error: could not create -darwin20-fake.rb
make: *** [-darwin20-fake.rb] Error 64
```

and perhap is just failing in some final symlinking step?

**Comment by bradical** [comment]
@ThomasKoppensteiner i saw you mentioned you got this working: https://github.com/rbenv/ruby-build/issues/1700#issuecomment-799541819. Were you on M1 Silicon or Intel?

@stevenjohn any luck on your end since: https://github.com/rbenv/ruby-build/issues/1700#issuecomment-761956133?

**Comment by ThomasKoppensteiner** [comment]
@bradical I'm using a Intel Core i7.

**Comment by asl331** [comment]
> I know that this is a rbenv post. But I want to comment that I was able to install an old Ruby (2.3.8) on my M1 with Big Sur 11.1, running XCode 12.3 with RVM using the following commands:
> 
> ```
> rvm pkg install openssl
> export optflags="-Wno-error=implicit-function-declaration"
> rvm install ruby-2.3.8 --with-openssl-dir=$HOME/.rvm/usr --rubygems ignore
> ```
> 
> 1. This install old openSSL as independent dependency (not used for the system)
> 2. Change a flag that (I think) was changed on XCode 12
> 3. Install  Ruby using the openSSL dependency from RVM (not from the system)

thanks a lot, worked for me.

**Comment by namratasaun** [comment]
> I know that this is a rbenv post. But I want to comment that I was able to install an old Ruby (2.3.8) on my M1 with Big Sur 11.1, running XCode 12.3 with RVM using the following commands:
> 
> ```
> rvm pkg install openssl
> export optflags="-Wno-error=implicit-function-declaration"
> rvm install ruby-2.3.8 --with-openssl-dir=$HOME/.rvm/usr --rubygems ignore
> ```
> 
> 1. This install old openSSL as independent dependency (not used for the system)
> 2. Change a flag that (I think) was changed on XCode 12
> 3. Install  Ruby using the openSSL dependency from RVM (not from the system)

you are doing this in Rosetta terminal or normal? @eveevans

**Comment by namratasaun** [comment]
![image](https://user-images.githubusercontent.com/35791871/128182213-988390d5-9f85-4ae7-bfbe-8e462a526fb1.png)
I got this as an error.

When I look into the log file, I get

![image](https://user-images.githubusercontent.com/35791871/128182562-a7a665d9-a826-46b6-af29-5293dc119553.png)

**Comment by namratasaun** [comment]
> I know that this is a rbenv post. But I want to comment that I was able to install an old Ruby (2.3.8) on my M1 with Big Sur 11.1, running XCode 12.3 with RVM using the following commands:
> 
> ```
> rvm pkg install openssl
> export optflags="-Wno-error=implicit-function-declaration"
> rvm install ruby-2.3.8 --with-openssl-dir=$HOME/.rvm/usr --rubygems ignore
> ```
> 
> 1. This install old openSSL as independent dependency (not used for the system)
> 2. Change a flag that (I think) was changed on XCode 12
> 3. Install  Ruby using the openSSL dependency from RVM (not from the system)

@eveevans I am having issues installing it with openssl. Later when It requires openssl it gives me error saying reinstall using openssl

**Comment by crgc** [comment]
The following worked for me:

`CFLAGS="-Wno-error=implicit-function-declaration" RUBY_CONFIGURE_OPTS='--with-readline-dir=/usr/local/opt/readline/' arch -x86_64 rbenv install 2.2.2`

**Comment by bradical** [comment]
is the bottom line that it's not possible to build 2.2.2 for arm64?

**Comment by crgc** [comment]
Hi @bradical - the above worked for me with 2.2.2 for arm64. For context, I'm on an M1 Macbook running Big Sur 11.5.2.

**Comment by bradical** [comment]
But if i'm understanding your code you're actually building for x86 using rosetta, no?

```
 arch -x86_64
```

**Comment by crgc** [comment]
Yes, you're right. It's the only way I could make it work, at least with the older Rubies.

**Comment by smudge** [comment]
> This is interesting. It seems to have gotten pretty far and perhaps is just failing in some final symlinking step due to a badly named file -darwin20-fake.rb?

On Ruby 2.6 I'm seeing `RUBY_PLATFORM` return `-darwin21` instead of `arm64-darwin21` -- it's otherwise building successfully with `arm64` (after I add `RUBY_CFLAGS=-DUSE_FFI_CLOSURE_ALLOC`):

```bash
$ file $(rbenv which ruby)
[...]/.rbenv/versions/2.6.5/bin/ruby: Mach-O 64-bit executable arm64
```

But then:

```bash
$ ruby -e "puts RUBY_PLATFORM"
-darwin21
```

So the the "badly named file" (`-darwin20-fake.rb`) is likely because the architecture is showing up blank somewhere and leads to a platform string like `-darwin20`. Anyone know how that string gets set and if it can be fixed at build time (or at runtime even)?

**Comment by bradical** [comment]
> Yes, you're right. It's the only way I could make it work, at least with the older Rubies.

@crgc can you confirm the versions of readline and openssl that you have installed and also your macOS version? I'm trying your command https://github.com/rbenv/ruby-build/issues/1700#issuecomment-986584406 but it's not working for me.
